pkg/disks/loop.go

```diff
@@ -0,0 +1,235 @@
+	}
+
+	klog.V(4).InfoS("Creating loop device image", "Image", imagePath, "Size", sizeBytes)
+	args := []string{
+		fmt.Sprintf("if=%s", path.Join(devfsPath, "zero")),
+		fmt.Sprintf("of=%s", imagePath),
+		"bs=1",
+		"count=0",
+		fmt.Sprintf("seek=%d", sizeBytes),
+	}
+
+	stdout, stderr, err := oexec.RunCommand(ctx, executor, "dd", args...)
+	if err != nil {
+		return false, fmt.Errorf("can't run dd with args %v: %w, stdout: %q, stderr: %q", args, err, stdout.String(), stderr.String())
+	}
+
+	return true, nil
+}
+
+func attachLoopDevice(ctx context.Context, executor exec.Interface, imagePath string) (bool, string, error) {
+	loopDevices, err := losetup.ListLoopDevices(ctx, executor)
+	if err != nil {
+		return false, "", fmt.Errorf("can't list loop devices: %w", err)
+	}
+
+	for _, ld := range loopDevices {
+		if imagePath == ld.BackingFile {
+			klog.V(4).InfoS("Loop device already exists, nothing to do.", "Device", ld.Name, "Image", ld.BackingFile)
+			return false, ld.Name, nil
+		}
+	}
+
+	klog.V(4).InfoS("Attaching loop device to image.", "Image", imagePath)
+	devicePath, err := losetup.AttachLoopDevice(ctx, executor, imagePath)
+	if err != nil {
+		return false, "", fmt.Errorf("can't attach loop device to image %q: %w", imagePath, err)
+	}
+	klog.V(4).InfoS("Successfully attached loop device to image.", "Device", devicePath, "Image", imagePath)
+
+	return true, devicePath, nil
+}
+
+func makeDeviceSymlink(loopDevicePath, devicePath string) (bool, error) {
+	symlinkDirPath := path.Dir(devicePath)
+	_, err := os.Stat(symlinkDirPath)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return false, fmt.Errorf("can't stat %q: %w", symlinkDirPath, err)
+	}
+
+	if errors.Is(err, fs.ErrNotExist) {
+		err = os.Mkdir(symlinkDirPath, 0770)
+		if err != nil {
+			return false, fmt.Errorf("can't mkdir %q: %w", symlinkDirPath, err)
+		}
+	}
+
+	_, err = os.Stat(devicePath)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return false, fmt.Errorf("can't stat %q: %w", devicePath, err)
+	}
+
+	if errors.Is(err, fs.ErrNotExist) {
+		klog.V(4).InfoS("Creating device symlink to loop device", "Device", loopDevicePath, "Symlink", devicePath)
+		err = os.Symlink(loopDevicePath, devicePath)
+		if err != nil {
+			return false, fmt.Errorf("can't create symlink from %q to %q: %w", loopDevicePath, devicePath, err)
+		}
+
+		return true, nil
+	}
+
+	klog.V(4).InfoS("Device symlink already exists, nothing to do.", "Device", loopDevicePath, "Symlink", devicePath)
+	return false, nil
+}
+
+func deleteSymlinkToDevice(deviceSymlinkPath string) error {
+	klog.V(4).InfoS("Removing device symlink", "Symlink", deviceSymlinkPath)
+	err := os.Remove(deviceSymlinkPath)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return fmt.Errorf("can't remove device symlink %q: %w", deviceSymlinkPath, err)
+	}
+	if errors.Is(err, fs.ErrNotExist) {
+		klog.V(4).InfoS("Device symlink does not exist. Nothing to do.", "DeviceSymlink", deviceSymlinkPath)
```

Comment:
to be future proof, this is a good place to return

Reply:
done